File: go/consumer.go
// Consumer handles a long-lived subscription to a streamq topic.
//
// Design:
// The consumer exposes a single read channel (Messages()) and manages all
// connection lifecycle internally, initial connect, reconnect on failure,
// clean shutdown. The caller's code is a simple range loop:
//
//   for msg := range consumer.Messages() {
//       process(msg)
//       consumer.Ack(msg.Offset) // for WebSocket + group consumers
//   }
//   if err := consumer.Err(); err != nil {
//       // connection failed permanently
//   }
//
// Reconnect behaviour:
// If the broker connection drops (network error, broker restart), the consumer
// waits ReconnectDelay and tries again. If the subscription had a group and
// a committed offset, the broker will resume from the committed offset
// automatically, no messages are redelivered. If MaxReconnectAttempts is
// set and exhausted, Messages() is closed and Err() returns ErrMaxReconnects.
//
// At-least-once delivery (WebSocket + group):
// The consumer does NOT auto-ack. Call Ack(offset) after successfully
// processing a message. Until acked, the broker considers the message
// undelivered and will redeliver from the last committed offset on reconnect.
//
// At-most-once delivery (SSE or ungrouped):
// No ack mechanism, the broker auto-commits on send. If the consumer
// crashes after receiving but before processing, the message is lost.
//
// Thread safety:
// Messages() and Err() are safe to call from any goroutine.
// Ack() is safe to call from any goroutine.
// Close() is safe to call from any goroutine and is idempotent.

package streamq

import (
	"bufio"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/gorilla/websocket"
)

// Consumer is a long-lived subscription to a streamq topic.
// Obtain one via client.Subscribe().
type Consumer struct {
	client *Client
	topic  string
	opts   SubscribeOptions
	ctx    context.Context
	cancel context.CancelFunc

	// msgs is the channel the caller reads from.
	// Closed when the consumer stops (either via Close() or fatal error).
	msgs chan Message

	// ackCh carries ack offsets from Ack() to the write goroutine (WS only).
	// Buffered so Ack() never blocks the caller.
	ackCh chan int64

	// err holds the terminal error (if any) after msgs is closed.
	// Protected by errMu.
	err   error
	errMu sync.RWMutex

	// closeOnce ensures Close() is idempotent.
	closeOnce sync.Once
}

// newConsumer creates a Consumer but does not start the connection.
// Call start() to begin.
func newConsumer(ctx context.Context, client *Client, topic string, opts SubscribeOptions) *Consumer {
	ctx, cancel := context.WithCancel(ctx)
	return &Consumer{
		client: client,
		topic:  topic,
		opts:   opts,
		ctx:    ctx,
		cancel: cancel,
		msgs:   make(chan Message, opts.BufferSize),
		ackCh:  make(chan int64, opts.BufferSize),
	}
}

// start launches the background goroutine that manages the connection.
func (c *Consumer) start() {
	go c.run()
}

// Messages returns the channel of incoming messages.
// The channel is closed when the consumer stops.
// Range over this channel:
//
//	for msg := range consumer.Messages() { ... }
func (c *Consumer) Messages() <-chan Message {
	return c.msgs
}

// Ack sends an acknowledgement for the given offset to the broker.
// Only meaningful for WebSocket consumers in a consumer group —
// the broker advances the group's committed offset on receipt.
//
// For SSE consumers or ungrouped consumers, Ack is a no-op.
// Safe to call from any goroutine.
func (c *Consumer) Ack(offset int64) {
	if c.opts.Protocol == ProtocolSSE || c.opts.Group == "" {
		return
	}
	select {
	case c.ackCh <- offset:
	default:
		// ack channel full, this shouldn't happen under normal operation
		// (caller is acking faster than the write goroutine can send).
		// Drop silently; the worst case is the broker redelivers on reconnect.
	}
}

// Err returns the terminal error that caused the consumer to stop.
// Returns nil if the consumer was closed cleanly via Close(), or if it
// is still running.
// Call after the Messages() channel is closed to distinguish clean
// shutdown from failure.
func (c *Consumer) Err() error {
	c.errMu.RLock()
	defer c.errMu.RUnlock()
	return c.err
}

// Close stops the consumer and closes the Messages() channel.
// Idempotent, safe to call multiple times.
func (c *Consumer) Close() {
	c.closeOnce.Do(func() {
		c.cancel()
	})
}

// run is the main consumer loop. It manages reconnects and delegates to
// the protocol-specific connect function.
func (c *Consumer) run() {
	defer close(c.msgs)

	attempts := 0
	for {
		// Check if we've been closed before attempting a connection.
		select {
		case <-c.ctx.Done():
			// Clean close, set ErrConsumerClosed so callers know it was intentional.
			c.setErr(&ErrConsumerClosed{})
			return
		default:
		}

		// Connect using the selected protocol.
		var connErr error
		switch c.opts.Protocol {
		case ProtocolSSE:
			connErr = c.connectSSE()
		default:
			connErr = c.connectWebSocket()
		}

		// nil error means the context was cancelled (clean close).
		if connErr == nil {
			c.setErr(&ErrConsumerClosed{})
			return
		}

		// Context cancelled during connection, clean shutdown.
		if c.ctx.Err() != nil {
			c.setErr(&ErrConsumerClosed{})
			return
		}

		attempts++

		// Check reconnect limit.
		if c.client.opts.MaxReconnectAttempts > 0 && attempts >= c.client.opts.MaxReconnectAttempts {
			c.setErr(&ErrMaxReconnects{Attempts: attempts, Last: connErr})
			return
		}

		// Wait before reconnecting. Respect context cancellation during wait.
		select {
		case <-time.After(c.client.opts.ReconnectDelay):
		case <-c.ctx.Done():
			c.setErr(&ErrConsumerClosed{})
			return
		}
	}
}

// connectWebSocket establishes a WebSocket connection and pumps messages
// until the connection drops or the context is cancelled.
// Returns nil on clean shutdown, error on connection failure.
func (c *Consumer) connectWebSocket() error {
	wsURL := c.buildURL("ws")

	conn, _, err := websocket.DefaultDialer.DialContext(c.ctx, wsURL, nil)
	if err != nil {
		return fmt.Errorf("ws dial: %w", err)
	}
	defer conn.Close()

	// done is closed when the read pump exits, signalling the write pump to stop.
	done := make(chan struct{})

	// Write pump: sends ack frames to the broker.
	// Runs in a separate goroutine, gorilla/websocket requires exactly one
	// concurrent writer and one concurrent reader.
	go func() {
		defer conn.Close()
		for {
			select {
			case offset := <-c.ackCh:
				frame := ackFrame{Ack: offset}
				data, err := json.Marshal(frame)
				if err != nil {
					continue
				}
				if err := conn.WriteMessage(websocket.TextMessage, data); err != nil {
					return
				}
			case <-done:
				conn.WriteMessage(
					websocket.CloseMessage,
					websocket.FormatCloseMessage(websocket.CloseNormalClosure, ""),
				)
				return
			case <-c.ctx.Done():
				conn.WriteMessage(
					websocket.CloseMessage,
					websocket.FormatCloseMessage(websocket.CloseNormalClosure, ""),
				)
				return
			}
		}
	}()

	// Read pump: receives message frames from the broker.
	defer close(done)
	for {
		_, data, err := conn.ReadMessage()
		if err != nil {
			// Distinguish clean close from network error.
			if websocket.IsCloseError(err, websocket.CloseNormalClosure, websocket.CloseGoingAway) {
				return nil
			}
			if c.ctx.Err() != nil {
				return nil
			}
			return fmt.Errorf("ws read: %w", err)
		}

		var msg Message
		if err := json.Unmarshal(data, &msg); err != nil {
			// Malformed frame, skip.
			continue
		}

		// Send to caller's channel. Respect context cancellation.
		select {
		case c.msgs <- msg:
		case <-c.ctx.Done():
			return nil
		}
	}
}

// connectSSE establishes an SSE connection and pumps events until the
// connection drops or the context is cancelled.
// Returns nil on clean shutdown, error on connection failure.
func (c *Consumer) connectSSE() error {
	sseURL := c.buildURL("sse")

	req, err := http.NewRequestWithContext(c.ctx, http.MethodGet, sseURL, nil)
	if err != nil {
		return fmt.Errorf("sse: build request: %w", err)
	}
	req.Header.Set("Accept", "text/event-stream")
	req.Header.Set("Cache-Control", "no-cache")

	// Use a client without a timeout for the SSE connection —
	// the base client's Timeout applies to the entire response body,
	// which would terminate the stream prematurely.
	httpClient := &http.Client{}
	resp, err := httpClient.Do(req)
	if err != nil {
		if c.ctx.Err() != nil {
			return nil
		}
		return fmt.Errorf("sse: connect: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("sse: unexpected status %d", resp.StatusCode)
	}

	// Parse SSE events line by line.
	// SSE format:
	//   id: <offset>
	//   data: <json>
	//   <blank line>  ← event boundary
	scanner := bufio.NewScanner(resp.Body)
	var dataLine string

	for scanner.Scan() {
		line := scanner.Text()

		if strings.HasPrefix(line, "data: ") {
			dataLine = strings.TrimPrefix(line, "data: ")
			continue
		}

		// Blank line = end of event. Process if we have data.
		if line == "" && dataLine != "" {
			var msg Message
			if err := json.Unmarshal([]byte(dataLine), &msg); err != nil {
				dataLine = ""
				continue
			}

			select {
			case c.msgs <- msg:
			case <-c.ctx.Done():
				return nil
			}

			dataLine = ""
		}
	}

	if err := scanner.Err(); err != nil {
		if c.ctx.Err() != nil {
			return nil
		}
		return fmt.Errorf("sse: read: %w", err)
	}

	return nil
}

// buildURL constructs the subscription URL for the given protocol ("ws" or "sse").
//
// WebSocket:  ws://host/topics/<name>/subscribe/ws?group=X&from_offset=Y
// SSE:        http://host/topics/<name>/subscribe/sse?group=X&from_offset=Y
func (c *Consumer) buildURL(protocol string) string {
	base := c.client.baseURL

	var scheme string
	switch protocol {
	case "ws":
		// Replace http(s) with ws(s) for WebSocket URLs.
		scheme = strings.Replace(base, "https://", "wss://", 1)
		scheme = strings.Replace(scheme, "http://", "ws://", 1)
	default:
		scheme = base
	}

	path := fmt.Sprintf("%s/topics/%s/subscribe/%s", scheme, c.topic, protocol)

	params := url.Values{}
	if c.opts.Group != "" {
		params.Set("group", c.opts.Group)
	}
	if c.opts.FromOffset >= 0 {
		params.Set("from_offset", strconv.FormatInt(c.opts.FromOffset, 10))
	}

	if len(params) > 0 {
		return path + "?" + params.Encode()
	}
	return path
}

// setErr stores the terminal error.
// Only the first call has effect, errors don't overwrite each other.
func (c *Consumer) setErr(err error) {
	c.errMu.Lock()
	defer c.errMu.Unlock()
	if c.err == nil {
		c.err = err
	}
}

File: python/streamq/consumer.py
# Async consumer for streamq topics.
#
# Design — async generator pattern:
# The consumer exposes an async generator via __aiter__ / __anext__.
# This is the most Pythonic way to consume a stream:
#
#   async for msg in consumer:
#       await process(msg)
#       await consumer.ack(msg.offset)
#
# Why async generator instead of asyncio.Queue?
# A queue-based API requires the caller to manage a background task and
# remember to cancel it. The async generator pattern handles all lifecycle
# internally — the generator runs until the connection drops, then
# reconnects transparently. Cleanup happens in the finally block.
#
# Reconnect behaviour:
# On connection drop, the consumer waits reconnect_delay seconds and retries.
# For grouped consumers with a committed offset, the broker resumes from
# where the group left off automatically. If max_reconnect_attempts is
# exceeded, StreamqMaxReconnects is raised from the generator.
#
# Protocol support:
# - WebSocket: bidirectional, supports explicit ack() for at-least-once delivery
# - SSE: unidirectional, at-most-once (no ack channel)
#
# Thread safety:
# The consumer is not thread-safe. Use it from a single asyncio event loop.
# For multi-threaded use, see SyncConsumer in sync.py.

from __future__ import annotations

import asyncio
import json
import logging
from typing import AsyncIterator, Optional
from urllib.parse import urlencode

import aiohttp

from .exceptions import StreamqConsumerClosed, StreamqMaxReconnects
from .models import Message

logger = logging.getLogger(__name__)


class AsyncConsumer:
    """Async consumer for a streamq topic.

    Obtain via AsyncClient.subscribe(). Use as an async context manager
    or async iterator:

        async with client.subscribe("payments") as consumer:
            async for msg in consumer:
                print(msg.payload)
                await consumer.ack(msg.offset)

    Or without context manager:

        consumer = client.subscribe("payments")
        async for msg in consumer:
            ...
        await consumer.close()
    """

    def __init__(
        self,
        base_url: str,
        topic: str,
        *,
        group: str = "",
        from_offset: int = -1,
        protocol: str = "ws",  # "ws" or "sse"
        buffer_size: int = 64,
        reconnect_delay: float = 2.0,
        max_reconnect_attempts: int = 0,  # 0 = unlimited
        session: Optional[aiohttp.ClientSession] = None,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._topic = topic
        self._group = group
        self._from_offset = from_offset
        self._protocol = protocol
        self._buffer_size = buffer_size
        self._reconnect_delay = reconnect_delay
        self._max_reconnect_attempts = max_reconnect_attempts

        # If a session is passed in (from AsyncClient), we don't own it
        # and should not close it. If we create our own, we close it on exit.
        self._session = session
        self._owns_session = session is None

        self._closed = False
        self._close_event = asyncio.Event()

        # ack_queue carries offset values from ack() to the WebSocket write loop.
        # Buffered so ack() never blocks.
        self._ack_queue: asyncio.Queue[int] = asyncio.Queue(maxsize=buffer_size)

    async def __aenter__(self) -> "AsyncConsumer":
        return self

    async def __aexit__(self, *_) -> None:
        await self.close()

    def __aiter__(self) -> AsyncIterator[Message]:
        return self._stream()

    async def _stream(self) -> AsyncIterator[Message]:
        """Main generator — connects, yields messages, reconnects on failure."""
        if self._owns_session:
            self._session = aiohttp.ClientSession()

        attempts = 0
        last_error: Optional[Exception] = None

        try:
            while not self._closed:
                try:
                    if self._protocol == "sse":
                        async for msg in self._connect_sse():
                            yield msg
                    else:
                        async for msg in self._connect_ws():
                            yield msg

                    # Generator exhausted cleanly (broker closed the connection).
                    # If we weren't explicitly closed, reconnect.
                    if self._closed:
                        return

                except StreamqConsumerClosed:
                    return

                except asyncio.CancelledError:
                    return

                except Exception as e:
                    last_error = e
                    attempts += 1
                    logger.warning(
                        "streamq consumer: connection lost (%s), attempt %d",
                        e,
                        attempts,
                    )

                    if (
                        self._max_reconnect_attempts > 0
                        and attempts >= self._max_reconnect_attempts
                    ):
                        raise StreamqMaxReconnects(attempts, last_error)

                    # Wait before reconnecting. Respect close() during wait.
                    try:
                        await asyncio.wait_for(
                            self._close_event.wait(),
                            timeout=self._reconnect_delay,
                        )
                        # close_event fired — stop.
                        return
                    except asyncio.TimeoutError:
                        pass  # reconnect delay elapsed — try again

        finally:
            if self._owns_session and self._session:
                await self._session.close()
                self._session = None

    async def _connect_ws(self) -> AsyncIterator[Message]:
        """Connect via WebSocket and yield messages until disconnected."""
        url = self._build_url("ws")
        logger.debug("streamq: connecting WebSocket to %s", url)

        async with self._session.ws_connect(url) as ws:
            # Launch write pump as a task, sends ack frames to the broker.
            # We use a task (not gather) so the read loop can yield independently.
            write_task = asyncio.create_task(self._ws_write_pump(ws))

            try:
                async for raw in ws:
                    if raw.type == aiohttp.WSMsgType.TEXT:
                        try:
                            data = json.loads(raw.data)
                            msg = Message.from_dict(data)
                            yield msg
                        except (json.JSONDecodeError, KeyError) as e:
                            logger.warning("streamq: malformed ws frame: %s", e)
                            continue

                    elif raw.type == aiohttp.WSMsgType.CLOSE:
                        break

                    elif raw.type == aiohttp.WSMsgType.ERROR:
                        raise aiohttp.ClientError(f"ws error: {ws.exception()}")
            finally:
                write_task.cancel()
                try:
                    await write_task
                except asyncio.CancelledError:
                    pass

    async def _ws_write_pump(self, ws: aiohttp.ClientWebSocketResponse) -> None:
        """Send ack frames from the ack queue to the broker.

        Runs as a background task while the read loop is active.
        Cancelled when the read loop exits.
        """
        while True:
            offset = await self._ack_queue.get()
            frame = json.dumps({"ack": offset})
            try:
                await ws.send_str(frame)
            except Exception as e:
                logger.warning(
                    "streamq: failed to send ack for offset %d: %s", offset, e
                )

    async def _connect_sse(self) -> AsyncIterator[Message]:
        """Connect via SSE and yield messages until disconnected."""
        url = self._build_url("sse")
        logger.debug("streamq: connecting SSE to %s", url)

        headers = {
            "Accept": "text/event-stream",
            "Cache-Control": "no-cache",
        }

        async with self._session.get(url, headers=headers) as resp:
            if resp.status != 200:
                raise aiohttp.ClientResponseError(
                    resp.request_info,
                    resp.history,
                    status=resp.status,
                )

            # Parse SSE line by line.
            # SSE format:
            #   data: <json>
            #   <blank line>  ← event boundary
            data_line = ""

            async for line_bytes in resp.content:
                if self._closed:
                    return

                line = line_bytes.decode("utf-8").rstrip("\n\r")

                if line.startswith("data: "):
                    data_line = line[6:]  # strip "data: " prefix

                elif line == "" and data_line:
                    # Blank line = end of event.
                    try:
                        data = json.loads(data_line)
                        msg = Message.from_dict(data)
                        yield msg
                    except (json.JSONDecodeError, KeyError) as e:
                        logger.warning("streamq: malformed sse event: %s", e)
                    finally:
                        data_line = ""

    async def ack(self, offset: int) -> None:
        """Acknowledge a message offset.

        Only meaningful for WebSocket consumers in a consumer group.
        The broker advances the group's committed offset on receipt,
        providing at-least-once delivery guarantees.

        For SSE consumers or ungrouped consumers, this is a no-op.
        """
        if self._protocol == "sse" or not self._group:
            return

        try:
            self._ack_queue.put_nowait(offset)
        except asyncio.QueueFull:
            # Ack queue full, shouldn't happen under normal operation.
            # Drop silently; worst case is redelivery on reconnect.
            logger.warning(
                "streamq: ack queue full, dropping ack for offset %d", offset
            )

    async def close(self) -> None:
        """Stop the consumer and release resources.

        Idempotent — safe to call multiple times.
        After close(), the async iterator will stop yielding.
        """
        if not self._closed:
            self._closed = True
            self._close_event.set()

    def _build_url(self, protocol: str) -> str:
        """Build the subscription URL for the given protocol.

        WebSocket: ws://host/topics/<name>/subscribe/ws?group=X&from_offset=Y
        SSE:       http://host/topics/<name>/subscribe/sse?group=X&from_offset=Y
        """
        base = self._base_url

        if protocol == "ws":
            # Replace http(s) scheme with ws(s) for WebSocket URLs.
            base = base.replace("https://", "wss://").replace("http://", "ws://")

        path = f"{base}/topics/{self._topic}/subscribe/{protocol}"

        params: dict[str, str] = {}
        if self._group:
            params["group"] = self._group
        if self._from_offset >= 0:
            params["from_offset"] = str(self._from_offset)

        if params:
            return f"{path}?{urlencode(params)}"
        return path
